### PR TITLE
Issue #2290: Update constraint generator and introduce fhir-ig-spl module

### PR DESCRIPTION
**Pull request by prb112**
This PR introduces the `fhir-ig-spl` module that includes artifacts from the Structured Product Label implementation guide:
http://build.fhir.org/ig/HL7/fhir-spl/branches/main/artifacts.html

The following changes where also made:

- Refactor `ConstraintGenerator` to make it more maintainable
- Updated `ConstraintGenerator` to handle DiscriminatorType.PROFILE per Issue #2290
- Updated `ConstraintGenerator` to handle `Uri` and `Code` pattern value constraints
- Updated `ConstraintGenerator` to provide additional trace output when using log level FINEST
- Updated `ProfileSupport` to eliminate redundant constraints specified on an element definition that references one or more profiles
- Updated `ResolveFunction` to calculate the correct "root resource" when resolving internal fragment references
- Updated `ConformsToFunction` with an optional second argument that can be used to optimize constraint evaluation when used during validation -- the validator already validates resources against resource asserted profiles
- Moved `getResourceNode`, `getRootResourceNode`, `isContained`, and `getResourceNodes` methods from `FHIRValidator` to `FHIRPathUtil`

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>
Signed-off-by: John T.E. Timm <johntimm@us.ibm.com>

### CODE REVIEW

File: conformance/fhir-ig-spl/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+3 - Updated #usagent to Organization/1234
+
+``` json 
+    "organization": {
+            "reference": "Organization/1234"
+        },
+        "participatingOrganization": {
+            "reference": "Organization/1234"
+        }
+```
```
Comment (prb112) [inline conformance/fhir-ig-spl/CHANGELOG.md:17]:
```suggestion

```

File: fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java [hunk 2]
```diff
@@ -8,6 +8,8 @@
 
 import static com.ibm.fhir.model.type.String.string;
 import static com.ibm.fhir.path.util.FHIRPathUtil.evaluatesToBoolean;
+import static com.ibm.fhir.path.util.FHIRPathUtil.getResourceNode;
```
Comment (prb112) [inline fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java:11]: copyright year

File: fhir-profile/src/test/java/com/ibm/fhir/profile/test/BPConstraintGeneratorTest.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/*
+ * (C) Copyright IBM Corp. 2019, 2021
```
Comment (prb112) [inline fhir-profile/src/test/java/com/ibm/fhir/profile/test/BPConstraintGeneratorTest.java:2]:
```suggestion
 * (C) Copyright IBM Corp. 2021
```

**Review by prb112 — COMMENTED** [review]
LGTM small copyright year changes towards the end 
Since I created this PR, I think you have to approve @JohnTimm

File: fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java (hunk 2 above)
Comment (JohnTimm) [reply fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java:11]: Updated

**Review by JohnTimm — APPROVED** [review]
LGTM

File: conformance/fhir-ig-spl/src/test/java/com/ibm/fhir/ig/us/spl/ExamplesValidationTest.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    public static boolean verify(Path p) {
+        boolean invalid = false;
+        try (InputStream in = new FileInputStream(p.toFile())) {
+            System.out.println("Path: " + p);
+            Resource r = FHIRParser.parser(Format.JSON).parse(in);
+            List<Issue> validate = FHIRValidator.validator().validate(r);
+            for (Issue issue : validate) {
+                if (IssueSeverity.ERROR.equals(issue.getSeverity()) || IssueSeverity.ERROR.equals(issue.getSeverity())) {
+                    System.out.println(issue);
+//                  System.out.println(r);
```
Comment (prb112) [inline conformance/fhir-ig-spl/src/test/java/com/ibm/fhir/ig/us/spl/ExamplesValidationTest.java:65]:
```suggestion
```

**Pull request merged**